File: src/file_reader.nim
import std/os
import std/times
import std/strutils
import pathutils
import ignore_rules

const CACHE_SIZE* = 256
const MAX_PATH_LENGTH* = 4096
const DEFAULT_MAX_LINES* = 1000

type
  FileReaderError* {.pure.} = enum
    Success
    NullPath
    FileNotFound
    PermissionDenied
    ReadFailed
    MemoryAlloc

  LineRange* = object
    startLine*: int
    endLine*: int
    totalLines*: int
    truncated*: int

  FileReaderResult* = object
    content*: string
    range*: LineRange
    error*: FileReaderError
    errorMessage*: string

  FileCacheEntry* = object
    key*: string
    readCount*: int
    mtime*: Time

var cacheTable*: array[CACHE_SIZE, FileCacheEntry]

proc cacheHash(path: string): int =
  ## 计算路径的哈希值（大小写不敏感）
  var h: uint = 0
  for c in path:
    h = h * 31'u + ord(c.toLowerAscii()).uint
  result = (h mod CACHE_SIZE.uint).int

proc cacheGet*(absolutePath: string): var FileCacheEntry =
  ## 在缓存中查找路径对应的缓存项
  ## 返回缓存项的引用，调用方通过 readCount == 0 / key.len == 0 判断未命中
  let idx = cacheHash(absolutePath)
  if cacheTable[idx].readCount > 0 and cacheTable[idx].key == absolutePath:
    return cacheTable[idx]
  # 未命中，返回该槽位（不做任何修改）
  return cacheTable[idx]

proc cacheSet*(absolutePath: string, mtime: Time, readCount: int) =
  ## 将文件路径及元信息写入缓存槽位
  let idx = cacheHash(absolutePath)
  cacheTable[idx].key = absolutePath
  cacheTable[idx].readCount = readCount
  cacheTable[idx].mtime = mtime

proc cacheInvalidate*(absolutePath: string) =
  ## 使指定路径的缓存失效（清空 readCount）
  let idx = cacheHash(absolutePath)
  if cacheTable[idx].key == absolutePath:
    cacheTable[idx].readCount = 0

proc getFileMtime*(path: string): Time =
  ## 获取文件的最后修改时间
  ## 如果文件不存在，返回 epoch 时间（0）
  try:
    result = getLastModificationTime(path)
  except:
    result = fromUnix(0)

proc countLines*(content: string): int =
  ## 统计字符串中的换行符数量
  for c in content:
    if c == '\n':
      inc result

proc parseLineRange*(requestedStart, requestedEnd: int): LineRange =
  ## 解析行范围参数
  ## - 如果 requestedStart <= 0，从第 1 行开始
  ## - 如果 requestedEnd <= 0，默认读取 DEFAULT_MAX_LINES 行
  ## - 如果 requestedEnd > 0 且 endLine < startLine，自动交换
  result.startLine = if requestedStart > 0: requestedStart else: 1
  if requestedEnd > 0:
    result.endLine = requestedEnd
  else:
    result.endLine = result.startLine + DEFAULT_MAX_LINES - 1
  if requestedEnd > 0 and result.endLine < result.startLine:
    swap(result.startLine, result.endLine)
  result.truncated = 0

proc formatContentWithLineNumbers*(content: string, range: LineRange, totalLines: int): string =
  ## 格式化文件内容为带行号的输出
  ## 格式："{lineNum} | {content}"
  ## 尾部添加文件统计信息
  var
    pos = 0
    currentLine = 1
    startPos = 0
    endPos = content.len

  # 定位到起始行
  while currentLine < range.startLine and pos < content.len:
    if content[pos] == '\n':
      inc currentLine
    inc pos
  startPos = pos

  # 定位到结束行
  while currentLine <= range.endLine and pos < content.len:
    if content[pos] == '\n':
      inc currentLine
    inc pos
  endPos = pos

  # 去掉末尾换行
  if endPos > startPos and endPos > 0 and content[endPos - 1] == '\n':
    dec endPos

  result = newStringOfCap(endPos - startPos + (range.endLine - range.startLine + 1) * 15 + 200)

  var outLineNum = range.startLine
  var outPos = startPos

  while outPos < endPos:
    result.add($outLineNum)
    result.add(" | ")
    while outPos < endPos and content[outPos] != '\n':
      result.add(content[outPos])
      inc outPos
    if outPos < endPos and content[outPos] == '\n':
      result.add('\n')
      inc outPos
    inc outLineNum

  result.add("\n\n")
  if range.endLine < totalLines:
    result.add("(Showing lines ")
    result.add($range.startLine)
    result.add("-")
    result.add($range.endLine)
    result.add(" of ")
    result.add($totalLines)
    result.add(" total. Use start_line=")
    result.add($(range.endLine + 1))
    result.add(" to continue reading.)")
  else:
    result.add("(File has ")
    result.add($totalLines)
    result.add(" lines total.)")

proc readFileRange*(path: string; startLine, endLine: int): FileReaderResult =
  ## 读取文件的指定行范围，返回带行号格式化的内容
  ## - path: 文件路径（相对或绝对）
  ## - startLine: 起始行号（从 1 开始，<=0 表示第 1 行）
  ## - endLine: 结束行号（<=0 表示 DEFAULT_MAX_LINES 行）
  ##
  ## 流程（对应 C 的 file_read）：
  ## 1. 参数验证 → ERROR_NULL_PATH
  ## 2. clineignore 检查 → ERROR_PERMISSION_DENIED
  ## 3. 路径解析 → ERROR_FILE_NOT_FOUND
  ## 4. 缓存查找 + mtime 检测
  ## 5. 重复读取警告
  ## 6. 文件读取 → ERROR_READ_FAILED
  ## 7. 行统计 + 范围解析
  ## 8. 格式化输出
  ## 9. 拼接警告 + 返回结果

  # 1. 参数验证
  if path.len == 0:
    result.error = FileReaderError.NullPath
    result.errorMessage = "Path parameter is required"
    return

  # 2. clineignore 检查
  if checkIgnorePath(path):
    result.error = FileReaderError.PermissionDenied
    result.errorMessage = "Access denied by .clineignore rules"
    return

  # 3. 路径解析
  let absolutePath = resolveWorkspacePath(path)
  if absolutePath.len == 0:
    result.error = FileReaderError.FileNotFound
    result.errorMessage = "Could not resolve path"
    return

  # 4. 缓存查找 + mtime 检测
  var cached = addr(cacheGet(absolutePath))
  if cached.readCount > 0:
    let currentMtime = getFileMtime(absolutePath)
    if currentMtime != cached.mtime:
      cached.readCount = 0

  cached = addr(cacheGet(absolutePath))

  # 5. 重复读取警告
  var dupWarning = ""
  if cached.readCount > 0:
    inc cached.readCount
    if cached.readCount >= 3:
      dupWarning = "[DUPLICATE READ] You have already read '" & path & "' " & $cached.readCount & " times in this conversation. The content has not changed since your last read. Please use the information you already have and proceed with your task.\n\n"
    elif cached.readCount == 2:
      dupWarning = "[File already read] The file '" & path & "' was already read earlier in this conversation. Returning content:\n\n"

  # 6. 文件读取
  var content: string
  try:
    content = readFile(absolutePath)
  except CatchableError:
    if not fileExists(absolutePath):
      result.error = FileReaderError.ReadFailed
      result.errorMessage = "Error reading file: File not found"
    elif getFilePermissions(absolutePath) == {}:
      result.error = FileReaderError.ReadFailed
      result.errorMessage = "Error reading file: Permission denied"
    else:
      result.error = FileReaderError.ReadFailed
      result.errorMessage = "Error reading file: Read failed"
    return

  # 7. 行统计 + 范围解析
  let totalLines = countLines(content)
  var range = parseLineRange(startLine, endLine)
  range.totalLines = totalLines

  # 边界裁剪
  if range.startLine > totalLines:
    range.startLine = totalLines
  if range.endLine > totalLines:
    range.endLine = totalLines

  # 写入缓存（首次读取）
  if cached.readCount == 0:
    let mtime = getFileMtime(absolutePath)
    cacheSet(absolutePath, mtime, 1)

  # 8. 格式化输出
  let formattedContent = formatContentWithLineNumbers(content, range, totalLines)

  # 9. 拼接警告 + 返回结果
  if dupWarning.len > 0:
    result.content = dupWarning & formattedContent
  else:
    result.content = formattedContent
  result.range = range
  result.error = FileReaderError.Success

File: core/src/file_reader.rs
use std::sync::Mutex;
use std::sync::OnceLock;
use std::time::SystemTime;

use crate::ignore_rules::check_ignore_path;
use crate::pathutils::resolve_workspace_path;

const CACHE_SIZE: usize = 256;
const DEFAULT_MAX_LINES: i32 = 1000;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FileReaderError {
    Success,
    NullPath,
    FileNotFound,
    PermissionDenied,
    ReadFailed,
}

#[derive(Debug, Clone, Copy)]
pub struct LineRange {
    pub start_line: i32,
    pub end_line: i32,
    pub total_lines: i32,
    pub truncated: i32,
}

#[derive(Debug, Clone)]
pub struct FileReaderResult {
    pub content: String,
    pub range: LineRange,
    pub error: FileReaderError,
    pub error_message: String,
}

#[derive(Debug, Clone)]
pub(crate) struct FileCacheEntry {
    pub(crate) key: String,
    pub(crate) read_count: u32,
    pub(crate) mtime: SystemTime,
}

static CACHE: OnceLock<Mutex<Vec<FileCacheEntry>>> = OnceLock::new();

fn cache() -> &'static Mutex<Vec<FileCacheEntry>> {
    CACHE.get_or_init(|| {
        let mut v = Vec::with_capacity(CACHE_SIZE);
        for _ in 0..CACHE_SIZE {
            v.push(FileCacheEntry {
                key: String::new(),
                read_count: 0,
                mtime: SystemTime::UNIX_EPOCH,
            });
        }
        Mutex::new(v)
    })
}

fn cache_hash(path: &str) -> usize {
    let mut h: u64 = 0;
    for c in path.chars() {
        let lower = c.to_ascii_lowercase() as u64;
        h = h.wrapping_mul(31).wrapping_add(lower);
    }
    (h % CACHE_SIZE as u64) as usize
}

pub(crate) fn cache_get(absolute_path: &str) -> Option<FileCacheEntry> {
    let lock = cache().lock().unwrap();
    let idx = cache_hash(absolute_path);
    let entry = &lock[idx];
    if entry.read_count > 0 && entry.key == absolute_path {
        Some(entry.clone())
    } else {
        None
    }
}

pub(crate) fn cache_set(absolute_path: &str, mtime: SystemTime, read_count: u32) {
    let mut lock = cache().lock().unwrap();
    let idx = cache_hash(absolute_path);
    lock[idx] = FileCacheEntry {
        key: absolute_path.to_string(),
        read_count,
        mtime,
    };
}

fn cache_increment(absolute_path: &str) -> u32 {
    let mut lock = cache().lock().unwrap();
    let idx = cache_hash(absolute_path);
    let entry = &mut lock[idx];
    if entry.read_count > 0 && entry.key == absolute_path {
        entry.read_count += 1;
        entry.read_count
    } else {
        0
    }
}

pub fn cache_invalidate(absolute_path: &str) {
    let mut lock = cache().lock().unwrap();
    let idx = cache_hash(absolute_path);
    if lock[idx].key == absolute_path {
        lock[idx].read_count = 0;
    }
}

fn get_file_mtime(path: &str) -> SystemTime {
    std::fs::metadata(path)
        .and_then(|m| m.modified())
        .unwrap_or(SystemTime::UNIX_EPOCH)
}

fn count_lines(content: &str) -> i32 {
    let mut count = 0;
    for c in content.chars() {
        if c == '\n' {
            count += 1;
        }
    }
    count
}

fn parse_line_range(requested_start: i32, requested_end: i32) -> LineRange {
    let start_line = if requested_start > 0 {
        requested_start
    } else {
        1
    };
    let end_line = if requested_end > 0 {
        requested_end
    } else {
        start_line + DEFAULT_MAX_LINES - 1
    };
    let (start_line, end_line) = if requested_end > 0 && end_line < start_line {
        (end_line, start_line)
    } else {
        (start_line, end_line)
    };
    LineRange {
        start_line,
        end_line,
        total_lines: 0,
        truncated: 0,
    }
}

fn format_content_with_line_numbers(content: &str, range: &LineRange, total_lines: i32) -> String {
    let bytes = content.as_bytes();
    let mut pos = 0;
    let mut current_line = 1;

    while current_line < range.start_line && pos < bytes.len() {
        if bytes[pos] == b'\n' {
            current_line += 1;
        }
        pos += 1;
    }
    let start_pos = pos;

    while current_line <= range.end_line && pos < bytes.len() {
        if bytes[pos] == b'\n' {
            current_line += 1;
        }
        pos += 1;
    }
    let mut end_pos = pos;
    if end_pos > start_pos && end_pos > 0 && bytes[end_pos - 1] == b'\n' {
        end_pos -= 1;
    }

    let estimated_cap =
        (end_pos - start_pos) + (range.end_line - range.start_line + 1) as usize * 15 + 200;
    let mut result = String::with_capacity(estimated_cap);

    let mut out_line_num = range.start_line;
    let mut out_pos = start_pos;

    while out_pos < end_pos {
        result.push_str(&out_line_num.to_string());
        result.push_str(" | ");
        while out_pos < end_pos && bytes[out_pos] != b'\n' {
            result.push(bytes[out_pos] as char);
            out_pos += 1;
        }
        if out_pos < end_pos && bytes[out_pos] == b'\n' {
            result.push('\n');
            out_pos += 1;
        }
        out_line_num += 1;
    }

    result.push_str("\n\n");
    if range.end_line < total_lines {
        result.push_str(&format!(
            "(Showing lines {}-{} of {} total. Use start_line={} to continue reading.)",
            range.start_line,
            range.end_line,
            total_lines,
            range.end_line + 1
        ));
    } else {
        result.push_str(&format!("(File has {} lines total.)", total_lines));
    }

    result
}

pub fn read_file_range(path: &str, start_line: i32, end_line: i32) -> FileReaderResult {
    if path.is_empty() {
        return FileReaderResult {
            content: String::new(),
            range: LineRange {
                start_line: 0,
                end_line: 0,
                total_lines: 0,
                truncated: 0,
            },
            error: FileReaderError::NullPath,
            error_message: "Path parameter is required".to_string(),
        };
    }

    if check_ignore_path(path) {
        return FileReaderResult {
            content: String::new(),
            range: LineRange {
                start_line: 0,
                end_line: 0,
                total_lines: 0,
                truncated: 0,
            },
            error: FileReaderError::PermissionDenied,
            error_message: "Access denied by .crownignore rules".to_string(),
        };
    }

    let absolute_path = resolve_workspace_path(path);
    if absolute_path.is_empty() {
        return FileReaderResult {
            content: String::new(),
            range: LineRange {
                start_line: 0,
                end_line: 0,
                total_lines: 0,
                truncated: 0,
            },
            error: FileReaderError::FileNotFound,
            error_message: "Could not resolve path".to_string(),
        };
    }

    let mut cached = cache_get(&absolute_path);
    if let Some(ref entry) = cached
        && entry.read_count > 0
    {
        let current_mtime = get_file_mtime(&absolute_path);
        if current_mtime != entry.mtime {
            cached = None;
        }
    }

    let mut dup_warning = String::new();
    if cached.is_some() {
        let new_count = cache_increment(&absolute_path);
        if new_count >= 3 {
            dup_warning = format!(
                "[DUPLICATE READ] You have already read '{}' {} times in this conversation. The content has not changed since your last read. Please use the information you already have and proceed with your task.\n\n",
                path, new_count
            );
        } else if new_count == 2 {
            dup_warning = format!(
                "[File already read] The file '{}' was already read earlier in this conversation. Returning content:\n\n",
                path
            );
        }
    }

    let content = match std::fs::read_to_string(&absolute_path) {
        Ok(c) => c,
        Err(e) => {
            let msg = match e.kind() {
                std::io::ErrorKind::NotFound => "Error reading file: File not found",
                std::io::ErrorKind::PermissionDenied => "Error reading file: Permission denied",
                _ => "Error reading file: Read failed",
            };
            return FileReaderResult {
                content: String::new(),
                range: LineRange {
                    start_line: 0,
                    end_line: 0,
                    total_lines: 0,
                    truncated: 0,
                },
                error: FileReaderError::ReadFailed,
                error_message: msg.to_string(),
            };
        }
    };

    let total_lines = count_lines(&content);
    let mut range = parse_line_range(start_line, end_line);
    range.total_lines = total_lines;

    if range.start_line > total_lines {
        range.start_line = total_lines;
    }
    if range.end_line > total_lines {
        range.end_line = total_lines;
    }

    if cached.is_none() {
        let mtime = get_file_mtime(&absolute_path);
        cache_set(&absolute_path, mtime, 1);
    }

    let formatted_content = format_content_with_line_numbers(&content, &range, total_lines);

    let final_content = if dup_warning.is_empty() {
        formatted_content
    } else {
        format!("{}{}", dup_warning, formatted_content)
    };

    FileReaderResult {
        content: final_content,
        range,
        error: FileReaderError::Success,
        error_message: String::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serial_test::serial;
    use std::io::Write;

    #[test]
    fn test_null_path() {
        let result = read_file_range("", 1, 10);
        assert_eq!(result.error, FileReaderError::NullPath);
        assert_eq!(result.error_message, "Path parameter is required");
    }

    #[test]
    fn test_file_not_found() {
        let result = read_file_range("/nonexistent_path_xyz/file.txt", 1, 10);
        assert_eq!(result.error, FileReaderError::ReadFailed);
    }

    #[test]
    fn test_read_failed_not_found() {
        let dir = std::env::temp_dir();
        let path = dir.join("test_read_failed_not_found.txt");
        let result = read_file_range(path.to_str().unwrap(), 1, 10);
        assert_eq!(result.error, FileReaderError::ReadFailed);
        assert!(result.error_message.contains("File not found"));
    }

    #[test]
    fn test_read_failed_permission_denied() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_read_permission_denied.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        writeln!(f, "content").unwrap();

        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            std::fs::set_permissions(&file_path, std::fs::Permissions::from_mode(0o000)).unwrap();
            let result = read_file_range(file_path.to_str().unwrap(), 1, 10);
            assert_eq!(result.error, FileReaderError::ReadFailed);
            assert!(result.error_message.contains("Permission denied"));
            std::fs::set_permissions(&file_path, std::fs::Permissions::from_mode(0o644)).unwrap();
        }

        #[cfg(not(unix))]
        {
            let _ = f;
        }
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_basic_line_number_format() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_basic_line_number.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        writeln!(f, "line1").unwrap();
        writeln!(f, "line2").unwrap();
        writeln!(f, "line3").unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 1, 3);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("1 | line1"));
        assert!(result.content.contains("2 | line2"));
        assert!(result.content.contains("3 | line3"));
        assert!(result.content.contains("(File has 3 lines total.)"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_line_range_subset() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_line_range_subset.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        for i in 1..=10 {
            writeln!(f, "line{}", i).unwrap();
        }
        let result = read_file_range(file_path.to_str().unwrap(), 3, 5);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(!result.content.contains("1 | line1"));
        assert!(result.content.contains("3 | line3"));
        assert!(result.content.contains("5 | line5"));
        assert!(!result.content.contains("6 | line6"));
        assert!(result.content.contains("(Showing lines 3-5 of 10 total."));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_range_auto_swap() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_range_swap.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        for i in 1..=15 {
            writeln!(f, "line{}", i).unwrap();
        }
        let result = read_file_range(file_path.to_str().unwrap(), 10, 3);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("3 | line3"));
        assert!(result.content.contains("10 | line10"));
        assert!(!result.content.contains("1 | line1"));
        assert!(!result.content.contains("2 | line2"));
        assert!(result.content.contains("(Showing lines 3-10 of 15 total."));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_range_beyond_eof() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_range_beyond_eof.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        writeln!(f, "only one line").unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 1, 100);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("1 | only one line"));
        assert!(result.content.contains("(File has 1 lines total.)"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_single_line_no_newline() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_single_line_nonl.txt");
        std::fs::write(&file_path, "hello").unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("(File has 0 lines total.)"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_single_line_with_newline() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_single_line_nl.txt");
        std::fs::write(&file_path, "hello\n").unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("1 | hello"));
        assert!(result.content.contains("(File has 1 lines total.)"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_duplicate_read_warning_second() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_dup_read_second.txt");
        std::fs::write(&file_path, "content\n").unwrap();
        let abs_path = resolve_workspace_path(file_path.to_str().unwrap());

        cache_set(&abs_path, get_file_mtime(&abs_path), 1);

        let result = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("[File already read]"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_duplicate_read_warning_third() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_dup_read_third.txt");
        std::fs::write(&file_path, "content\n").unwrap();
        let abs_path = resolve_workspace_path(file_path.to_str().unwrap());

        cache_set(&abs_path, get_file_mtime(&abs_path), 2);

        let result = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("[DUPLICATE READ]"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_cache_mtime_invalidation() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_cache_mtime_inval.txt");
        std::fs::write(&file_path, "original\n").unwrap();

        let old_mtime = SystemTime::UNIX_EPOCH;
        let abs_path = resolve_workspace_path(file_path.to_str().unwrap());
        cache_set(&abs_path, old_mtime, 5);

        let result = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(!result.content.contains("[DUPLICATE READ]"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_absolute_path() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_absolute_path.txt");
        std::fs::write(&file_path, "test\n").unwrap();
        let abs_path = file_path.to_str().unwrap();
        let result = read_file_range(abs_path, 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("1 | test"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_relative_path() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_relative_path.txt");
        std::fs::write(&file_path, "test\n").unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("1 | test"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_cache_first_read_no_warning() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_cache_first_read.txt");
        std::fs::write(&file_path, "content\n").unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(!result.content.contains("[File already read]"));
        assert!(!result.content.contains("[DUPLICATE READ]"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_start_line_zero_defaults_to_one() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_start_line_zero.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        writeln!(f, "line1").unwrap();
        writeln!(f, "line2").unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 0, 2);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("1 | line1"));
        assert!(result.content.contains("2 | line2"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_cache_get_set_roundtrip() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_cache_roundtrip.txt");
        let abs_path = resolve_workspace_path(file_path.to_str().unwrap());
        cache_set(&abs_path, SystemTime::UNIX_EPOCH, 42);
        let entry = cache_get(&abs_path);
        assert!(entry.is_some());
        assert_eq!(entry.unwrap().read_count, 42);
        cache_invalidate(&abs_path);
        assert!(cache_get(&abs_path).is_none());
    }

    #[test]
    fn test_cache_warning_three_reads_sequential() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_cache_three_reads.txt");
        std::fs::write(&file_path, "content\n").unwrap();
        let r1 = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(r1.error, FileReaderError::Success);
        assert!(!r1.content.contains("[File already read]"));
        assert!(!r1.content.contains("[DUPLICATE READ]"));
        let r2 = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(r2.error, FileReaderError::Success);
        assert!(r2.content.contains("[File already read]"));
        let r3 = read_file_range(file_path.to_str().unwrap(), 1, 1);
        assert_eq!(r3.error, FileReaderError::Success);
        assert!(r3.content.contains("[DUPLICATE READ]"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_large_file_reports_correct_total() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_large_file.txt");
        let mut content = String::new();
        for i in 1..=2000 {
            content.push_str(&format!("line{}\n", i));
        }
        std::fs::write(&file_path, &content).unwrap();
        let result = read_file_range(file_path.to_str().unwrap(), 1, 2000);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("(File has 2000 lines total.)"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    fn test_read_file_range_shows_showing_lines_message() {
        let dir = std::env::temp_dir();
        let file_path = dir.join("test_showing_lines.txt");
        let mut f = std::fs::File::create(&file_path).unwrap();
        for i in 1..=10 {
            writeln!(f, "line{}", i).unwrap();
        }
        let result = read_file_range(file_path.to_str().unwrap(), 2, 5);
        assert_eq!(result.error, FileReaderError::Success);
        assert!(result.content.contains("2 | line2"));
        assert!(result.content.contains("5 | line5"));
        assert!(!result.content.contains("1 | line1"));
        assert!(result.content.contains("(Showing lines 2-5 of 10 total."));
        assert!(result.content.contains("start_line=6"));
        let _ = std::fs::remove_file(&file_path);
    }

    #[test]
    #[serial]
    fn test_read_file_denied_by_crownignore() {
        crate::ignore_rules::reset_ignore_rules();
        let original_dir = std::env::current_dir().unwrap();
        let temp_dir = std::env::temp_dir().join("crown_test_reader_crownignore");
        let _ = std::fs::create_dir_all(&temp_dir);
        std::env::set_current_dir(&temp_dir).unwrap();
        let ignore_path = temp_dir.join(".crownignore");
        let mut f = std::fs::File::create(&ignore_path).unwrap();
        writeln!(f, "secret.txt").unwrap();
        f.flush().unwrap();
        drop(f);
        crate::ignore_rules::reset_ignore_rules();
        let result = read_file_range("secret.txt", 1, 10);
        assert_eq!(result.error, FileReaderError::PermissionDenied);
        assert_eq!(result.error_message, "Access denied by .crownignore rules");
        let _ = std::fs::remove_file(&ignore_path);
        std::env::set_current_dir(&original_dir).unwrap();
        let _ = std::fs::remove_dir_all(&temp_dir);
    }
}
